common/schemas/keyman-touch-layout/keyman-touch-layout.spec.json

```diff
@@ -0,0 +1,144 @@
+        "displayUnderlying": { "type": "boolean" }
+      },
+      "required": ["layer"],
+      "additionalProperties": false
+    },
+
+    "layers": {
+      "type": "array",
+      "items": { "$ref": "#/definitions/layer" }
+    },
+
+    "layer": {
+      "type": "object",
+      "properties": {
+        "id": { "$ref": "#/definitions/layer-id" },
+        "row": { "$ref": "#/definitions/rows" }
+      },
+      "required": ["id", "row"],
+      "additionalProperties": false
+    },
+
+    "layer-id": {
+      "type": "string",
+      "pattern": "^[a-zA-Z0-9_-]+$"
+    },
+
+    "rows": {
+      "type": "array",
+      "items": { "$ref": "#/definitions/row" }
+    },
+
+    "row": {
+      "type": "object",
+      "properties": {
+        "id": { "$ref": "#/definitions/row-id" },
+        "key": { "$ref": "#/definitions/keys" }
+      },
+      "required": ["id", "key"],
+      "additionalProperties": false
+    },
+
+    "row-id": {
+      "oneOf": [
+        {
+          "type": "integer",
+          "minimum": 0,
+          "maximum": 100
+        },
+        {
+          "type": "string",
+          "pattern": "^[0-9]+$"
+        }
+      ]
+    },
+
+    "keys": {
+      "type": "array",
+      "items": { "$ref": "#/definitions/key" }
+    },
+
+    "key": {
+      "type": "object",
+      "properties": {
+        "id": { "$ref": "#/definitions/key-id" },
+        "text": { "type": "string" },
+        "layer": { "$ref": "#/definitions/layer-id" },
+        "nextlayer": { "$ref": "#/definitions/layer-id" },
+        "fontsize": { "type": "string" },
+        "font": { "$ref": "#/definitions/font-spec" },
+        "dk": { "type": "string" },
+        "sp": { "$ref" : "#/definitions/numeric" },
+        "pad": { "$ref" : "#/definitions/numeric" },
+        "width": { "$ref" : "#/definitions/numeric" },
+        "sk": { "$ref": "#/definitions/subkeys" }
+      },
+      "additionalProperties": false
+    },
+
+    "numeric": {
+      "oneOf": [
+        { "type": "string", "pattern": "^[0-9]+(\\.?[0-9]*)$|^$" },
+        { "type": "number" }
+      ]
+    },
+
+    "key-id": {
+      "type": "string",
+      "pattern": "^[TKUtku]_[a-zA-Z0-9_]+$|^$"
+    },
+
+    "subkeys": {
+      "type": "array",
+      "items": { "$ref": "#/definitions/subkey" }
+    },
+
+    "subkey": {
+      "type": "object",
+      "properties": {
+        "id": { "$ref": "#/definitions/key-id" },
+        "text": { "type": "string" },
+        "layer": { "$ref": "#/definitions/layer-id" },
+        "nextlayer": { "$ref": "#/definitions/layer-id" },
+        "sp": { "$ref" : "#/definitions/numeric" },
+        "pad": { "$ref" : "#/definitions/numeric" },
+        "width": { "$ref" : "#/definitions/numeric" },
+        "fontsize": { "type": "string" },
+        "font": { "$ref": "#/definitions/font-spec" },
+        "dk": { "type": "string" }
+      },
+      "required": ["id"],
+      "additionalProperties": false
+    },
```

Comment:
KMW will completely ignore the `pad` property for subkeys - I just double-checked.

It actually can use `width`, though I'm not 100% sure that's fully intentional - it's definitely not the average use-case ...enough so that I'm not 100% confident it'll be handled properly.  I don't recall ever testing against a keyboard with a subkey specifying that property.

As a minor note, I see that there are only two thing keeping this (as-is) from being WET with "key":
1. Here: `"required": ["id"]`
2. In "key": `properties { "sk" }`.

Is there no way to declare a common "parent type" and have the two derive from that?  Admittedly, I'm no schema expert, and perhaps you have thoughts & reasons why we wouldn't want to do that anyway.

Reply:
Yes, `pad`, `width`, `font` and `fontsize` are all incompletely supported. See corresponding README.md documentation. This spec allows things that are not currently supported by KMW because it matches what can currently be built.

Reply:
While I could declare a parent common type between `key` and `sk`, I opted instead to use per-field types which made the duplication reasonably pain-free, at least for now. JSON Schema's model of combining is somewhat complicated to read and work with, which makes it much easier to make mistakes, in my experience.